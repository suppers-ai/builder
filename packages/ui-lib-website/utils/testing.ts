// Testing utilities for component quality assurance

/**
 * Accessibility testing utilities
 */
export class AccessibilityTester {
  /**
   * Check if element has proper ARIA attributes
   */
  static checkAriaAttributes(element: Element): string[] {
    const issues: string[] = [];

    // Check for interactive elements without proper roles
    const interactiveElements = ["button", "a", "input", "select", "textarea"];
    if (interactiveElements.includes(element.tagName.toLowerCase())) {
      if (!element.getAttribute("role") && element.tagName.toLowerCase() === "div") {
        issues.push("Interactive div element missing role attribute");
      }
    }

    // Check for images without alt text
    if (element.tagName.toLowerCase() === "img") {
      if (!element.getAttribute("alt")) {
        issues.push("Image missing alt attribute");
      }
    }

    // Check for form inputs without labels
    if (["input", "select", "textarea"].includes(element.tagName.toLowerCase())) {
      const id = element.getAttribute("id");
      const ariaLabel = element.getAttribute("aria-label");
      const ariaLabelledBy = element.getAttribute("aria-labelledby");

      if (!ariaLabel && !ariaLabelledBy) {
        if (!id || !document.querySelector(`label[for="${id}"]`)) {
          issues.push("Form input missing accessible label");
        }
      }
    }

    // Check for proper heading hierarchy
    if (element.tagName.match(/^H[1-6]$/)) {
      const level = parseInt(element.tagName.charAt(1));
      const previousHeadings = Array.from(document.querySelectorAll("h1, h2, h3, h4, h5, h6"))
        .filter((h) =>
          document.compareDocumentPosition(element) & Node.DOCUMENT_POSITION_FOLLOWING
        );

      if (previousHeadings.length > 0) {
        const lastHeading = previousHeadings[previousHeadings.length - 1];
        const lastLevel = parseInt(lastHeading.tagName.charAt(1));

        if (level > lastLevel + 1) {
          issues.push(`Heading level ${level} skips hierarchy (previous was ${lastLevel})`);
        }
      }
    }

    return issues;
  }

  /**
   * Check color contrast ratios
   */
  static checkColorContrast(element: Element): string[] {
    const issues: string[] = [];
    const styles = globalThis.getComputedStyle(element);
    const color = styles.color;
    const backgroundColor = styles.backgroundColor;

    // Simple contrast check (would need more sophisticated algorithm in production)
    if (color && backgroundColor && color !== "transparent" && backgroundColor !== "transparent") {
      const colorLuminance = this.getLuminance(color);
      const bgLuminance = this.getLuminance(backgroundColor);
      const contrast = this.getContrastRatio(colorLuminance, bgLuminance);

      if (contrast < 4.5) {
        issues.push(`Low color contrast ratio: ${contrast.toFixed(2)} (minimum 4.5)`);
      }
    }

    return issues;
  }

  /**
   * Check keyboard navigation
   */
  static checkKeyboardNavigation(container: Element): string[] {
    const issues: string[] = [];
    const focusableElements = container.querySelectorAll(
      'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])',
    );

    focusableElements.forEach((element, index) => {
      const tabIndex = element.getAttribute("tabindex");

      // Check for positive tabindex (anti-pattern)
      if (tabIndex && parseInt(tabIndex) > 0) {
        issues.push(
          `Element has positive tabindex (${tabIndex}), which can disrupt natural tab order`,
        );
      }

      // Check for focus indicators
      const styles = globalThis.getComputedStyle(element, ":focus");
      if (!styles.outline || styles.outline === "none") {
        if (!styles.boxShadow && !styles.border) {
          issues.push("Element lacks visible focus indicator");
        }
      }
    });

    return issues;
  }

  private static getLuminance(color: string): number {
    // Simplified luminance calculation
    const rgb = color.match(/\d+/g);
    if (!rgb) return 0;

    const [r, g, b] = rgb.map((c) => {
      const channel = parseInt(c) / 255;
      return channel <= 0.03928 ? channel / 12.92 : Math.pow((channel + 0.055) / 1.055, 2.4);
    });

    return 0.2126 * r + 0.7152 * g + 0.0722 * b;
  }

  private static getContrastRatio(lum1: number, lum2: number): number {
    const lighter = Math.max(lum1, lum2);
    const darker = Math.min(lum1, lum2);
    return (lighter + 0.05) / (darker + 0.05);
  }
}

/**
 * Component testing utilities
 */
export class ComponentTester {
  /**
   * Test component rendering
   */
  static testRender(component: any, props: any = {}): boolean {
    try {
      // In a real implementation, this would use a testing framework
      // For now, just check if component can be instantiated
      const instance = component(props);
      return instance !== null && instance !== undefined;
    } catch (error) {
      console.error("Component render test failed:", error);
      return false;
    }
  }

  /**
   * Test component props
   */
  static testProps(
    component: any,
    requiredProps: string[],
    optionalProps: string[] = [],
  ): string[] {
    const issues: string[] = [];

    // Test with missing required props
    try {
      component({});
      if (requiredProps.length > 0) {
        issues.push("Component should require props but accepts empty object");
      }
    } catch (error) {
      // Expected for required props
    }

    // Test with valid props
    const validProps: any = {};
    requiredProps.forEach((prop) => {
      validProps[prop] = "test-value";
    });

    try {
      const result = component(validProps);
      if (!result) {
        issues.push("Component fails to render with valid props");
      }
    } catch (error) {
      issues.push(`Component throws error with valid props: ${error.message}`);
    }

    return issues;
  }

  /**
   * Test component accessibility
   */
  static testAccessibility(element: Element): string[] {
    const issues: string[] = [];

    issues.push(...AccessibilityTester.checkAriaAttributes(element));
    issues.push(...AccessibilityTester.checkColorContrast(element));
    issues.push(...AccessibilityTester.checkKeyboardNavigation(element));

    return issues;
  }
}

/**
 * Performance testing utilities
 */
export class PerformanceTester {
  /**
   * Measure component render time
   */
  static measureRenderTime(componentFn: () => void): number {
    const start = performance.now();
    componentFn();
    const end = performance.now();
    return end - start;
  }

  /**
   * Test memory usage (simplified)
   */
  static testMemoryUsage(testFn: () => void): any {
    if ("memory" in performance) {
      const beforeMemory = (performance as any).memory.usedJSHeapSize;
      testFn();
      const afterMemory = (performance as any).memory.usedJSHeapSize;

      return {
        before: beforeMemory,
        after: afterMemory,
        difference: afterMemory - beforeMemory,
      };
    }

    return { message: "Memory measurement not available" };
  }

  /**
   * Test for memory leaks in component cleanup
   */
  static testMemoryLeaks(
    createComponent: () => any,
    destroyComponent: (component: any) => void,
  ): boolean {
    const iterations = 100;
    const components: any[] = [];

    // Create many components
    for (let i = 0; i < iterations; i++) {
      components.push(createComponent());
    }

    // Destroy all components
    components.forEach((component) => destroyComponent(component));

    // Force garbage collection if available
    if ("gc" in window) {
      (window as any).gc();
    }

    // In a real test, you'd measure memory usage here
    return true; // Simplified
  }
}

/**
 * Cross-browser compatibility testing
 */
export class BrowserCompatibilityTester {
  /**
   * Check browser feature support
   */
  static checkFeatureSupport(): Record<string, boolean> {
    return {
      cssGrid: CSS.supports("display", "grid"),
      cssFlexbox: CSS.supports("display", "flex"),
      cssCustomProperties: CSS.supports("--test", "value"),
      intersectionObserver: "IntersectionObserver" in window,
      mutationObserver: "MutationObserver" in window,
      promiseSupport: "Promise" in window,
      fetchAPI: "fetch" in window,
      webComponents: "customElements" in window,
      serviceWorker: "serviceWorker" in navigator,
      localStorage: (() => {
        try {
          localStorage.setItem("test", "test");
          localStorage.removeItem("test");
          return true;
        } catch {
          return false;
        }
      })(),
    };
  }

  /**
   * Get browser information
   */
  static getBrowserInfo(): Record<string, string> {
    const ua = navigator.userAgent;

    return {
      userAgent: ua,
      platform: navigator.platform,
      language: navigator.language,
      cookieEnabled: navigator.cookieEnabled.toString(),
      onLine: navigator.onLine.toString(),
      vendor: navigator.vendor || "unknown",
      browserName: this.detectBrowser(ua),
      browserVersion: this.detectBrowserVersion(ua),
    };
  }

  private static detectBrowser(ua: string): string {
    if (ua.includes("Chrome")) return "Chrome";
    if (ua.includes("Firefox")) return "Firefox";
    if (ua.includes("Safari") && !ua.includes("Chrome")) return "Safari";
    if (ua.includes("Edge")) return "Edge";
    if (ua.includes("Opera")) return "Opera";
    return "Unknown";
  }

  private static detectBrowserVersion(ua: string): string {
    const match = ua.match(/(?:Chrome|Firefox|Safari|Edge|Opera)\/(\d+\.?\d*)/);
    return match ? match[1] : "Unknown";
  }
}

/**
 * Automated testing suite
 */
export class AutomatedTestSuite {
  private tests: Array<{
    name: string;
    test: () => Promise<boolean>;
    category: "accessibility" | "performance" | "compatibility" | "functionality";
  }> = [];

  /**
   * Add a test to the suite
   */
  addTest(
    name: string,
    test: () => Promise<boolean>,
    category: "accessibility" | "performance" | "compatibility" | "functionality",
  ) {
    this.tests.push({ name, test, category });
  }

  /**
   * Run all tests
   */
  async runTests(): Promise<{
    passed: number;
    failed: number;
    results: Array<{ name: string; passed: boolean; error?: string; category: string }>;
  }> {
    const results = [];
    let passed = 0;
    let failed = 0;

    for (const { name, test, category } of this.tests) {
      try {
        const result = await test();
        if (result) {
          passed++;
          results.push({ name, passed: true, category });
        } else {
          failed++;
          results.push({ name, passed: false, category });
        }
      } catch (error) {
        failed++;
        results.push({
          name,
          passed: false,
          error: error instanceof Error ? error.message : "Unknown error",
          category,
        });
      }
    }

    return { passed, failed, results };
  }

  /**
   * Generate test report
   */
  generateReport(testResults: any): string {
    const { passed, failed, results } = testResults;
    const total = passed + failed;
    const passRate = ((passed / total) * 100).toFixed(1);

    let report = `# Component Library Test Report\n\n`;
    report += `**Total Tests:** ${total}\n`;
    report += `**Passed:** ${passed}\n`;
    report += `**Failed:** ${failed}\n`;
    report += `**Pass Rate:** ${passRate}%\n\n`;

    // Group by category
    const categories = ["accessibility", "performance", "compatibility", "functionality"];

    categories.forEach((category) => {
      const categoryTests = results.filter((r: any) => r.category === category);
      if (categoryTests.length === 0) return;

      report += `## ${category.charAt(0).toUpperCase() + category.slice(1)} Tests\n\n`;

      categoryTests.forEach((result: any) => {
        const status = result.passed ? "✅" : "❌";
        report += `${status} **${result.name}**\n`;
        if (result.error) {
          report += `   Error: ${result.error}\n`;
        }
        report += "\n";
      });
    });

    return report;
  }
}
