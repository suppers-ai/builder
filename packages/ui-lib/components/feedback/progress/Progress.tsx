import { BaseComponentProps, ColorProps, SizeProps } from "../../types.ts";

export interface ProgressProps extends BaseComponentProps, SizeProps, ColorProps {
  value?: number;
  max?: number;
  indeterminate?: boolean;
  striped?: boolean;
  animated?: boolean;
}

export function Progress({
  class: className = "",
  size = "md",
  color = "primary",
  value = 0,
  max = 100,
  indeterminate = false,
  striped = false,
  animated = false,
  id,
  ...props
}: ProgressProps) {
  const progressClasses = [
    "progress",
    color ? `progress-${color}` : "",
    size ? `progress-${size}` : "",
    striped ? "progress-striped" : "",
    animated ? "progress-animated" : "",
    className,
  ].filter(Boolean).join(" ");

  if (indeterminate) {
    return <progress class={progressClasses} id={id} {...props}></progress>;
  }

  return (
    <progress
      class={progressClasses}
      value={value}
      max={max}
      id={id}
      {...props}
    >
      {value}%
    </progress>
  );
}
