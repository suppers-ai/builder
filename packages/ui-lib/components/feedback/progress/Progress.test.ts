import { assertEquals, assertStringIncludes } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { renderToString } from "preact-render-to-string";
import { Progress } from "./Progress.tsx";

Deno.test("Progress - basic rendering", () => {
  const html = renderToString(Progress({}));
  assertStringIncludes(html, "progress");
  assertStringIncludes(html, "progress-primary");
  assertStringIncludes(html, "progress-md");
  assertStringIncludes(html, 'value="0"');
  assertStringIncludes(html, 'max="100"');
});

Deno.test("Progress - with value", () => {
  const html = renderToString(Progress({
    value: 50,
  }));
  assertStringIncludes(html, 'value="50"');
  assertStringIncludes(html, "50%");
});

Deno.test("Progress - with custom max", () => {
  const html = renderToString(Progress({
    value: 30,
    max: 200,
  }));
  assertStringIncludes(html, 'value="30"');
  assertStringIncludes(html, 'max="200"');
  assertStringIncludes(html, "30%");
});

Deno.test("Progress - indeterminate", () => {
  const html = renderToString(Progress({
    indeterminate: true,
  }));
  assertStringIncludes(html, "progress");
  assertEquals(html.includes("value="), false);
  assertEquals(html.includes("max="), false);
  assertEquals(html.includes("%"), false);
});

Deno.test("Progress - not indeterminate (default)", () => {
  const html = renderToString(Progress({
    indeterminate: false,
    value: 25,
  }));
  assertStringIncludes(html, 'value="25"');
  assertStringIncludes(html, "25%");
});

Deno.test("Progress - striped", () => {
  const html = renderToString(Progress({
    striped: true,
  }));
  assertStringIncludes(html, "progress-striped");
});

Deno.test("Progress - animated", () => {
  const html = renderToString(Progress({
    animated: true,
  }));
  assertStringIncludes(html, "progress-animated");
});

Deno.test("Progress - striped and animated", () => {
  const html = renderToString(Progress({
    striped: true,
    animated: true,
  }));
  assertStringIncludes(html, "progress-striped");
  assertStringIncludes(html, "progress-animated");
});

Deno.test("Progress - size variants", () => {
  const sizes = ["xs", "sm", "md", "lg"];

  sizes.forEach((size) => {
    const html = renderToString(Progress({
      size: size as 'xs' | 'sm' | 'md' | 'lg' | 'xl',
      value: 50,
    }));
    assertStringIncludes(html, `progress-${size}`);
  });
});

Deno.test("Progress - color variants", () => {
  const colors = ["primary", "secondary", "accent", "info", "success", "warning", "error"];

  colors.forEach((color) => {
    const html = renderToString(Progress({
      color: color as 'primary' | 'secondary' | 'accent' | 'neutral' | 'info' | 'success' | 'warning' | 'error',
      value: 60,
    }));
    assertStringIncludes(html, `progress-${color}`);
  });
});

Deno.test("Progress - custom class", () => {
  const html = renderToString(Progress({
    class: "custom-progress-class",
    value: 40,
  }));
  assertStringIncludes(html, "custom-progress-class");
});

Deno.test("Progress - with id", () => {
  const html = renderToString(Progress({
    id: "test-progress",
    value: 75,
  }));
  assertStringIncludes(html, 'id="test-progress"');
});

Deno.test("Progress - zero value", () => {
  const html = renderToString(Progress({
    value: 0,
  }));
  assertStringIncludes(html, 'value="0"');
  assertStringIncludes(html, "0%");
});

Deno.test("Progress - full value", () => {
  const html = renderToString(Progress({
    value: 100,
  }));
  assertStringIncludes(html, 'value="100"');
  assertStringIncludes(html, "100%");
});

Deno.test("Progress - over max value", () => {
  const html = renderToString(Progress({
    value: 150,
    max: 100,
  }));
  assertStringIncludes(html, 'value="150"');
  assertStringIncludes(html, 'max="100"');
  assertStringIncludes(html, "150%");
});

Deno.test("Progress - all props combined", () => {
  const html = renderToString(Progress({
    value: 85,
    max: 100,
    size: "lg",
    color: "success",
    striped: true,
    animated: true,
    class: "test-class",
    id: "full-progress",
  }));
  assertStringIncludes(html, 'value="85"');
  assertStringIncludes(html, 'max="100"');
  assertStringIncludes(html, "progress-lg");
  assertStringIncludes(html, "progress-success");
  assertStringIncludes(html, "progress-striped");
  assertStringIncludes(html, "progress-animated");
  assertStringIncludes(html, "test-class");
  assertStringIncludes(html, 'id="full-progress"');
  assertStringIncludes(html, "85%");
});

Deno.test("Progress - indeterminate with styles", () => {
  const html = renderToString(Progress({
    indeterminate: true,
    size: "lg",
    color: "warning",
    striped: true,
    animated: true,
  }));
  assertStringIncludes(html, "progress-lg");
  assertStringIncludes(html, "progress-warning");
  assertStringIncludes(html, "progress-striped");
  assertStringIncludes(html, "progress-animated");
  assertEquals(html.includes("value="), false);
  assertEquals(html.includes("%"), false);
});

Deno.test("Progress - small primary progress", () => {
  const html = renderToString(Progress({
    size: "sm",
    color: "primary",
    value: 25,
  }));
  assertStringIncludes(html, "progress-sm");
  assertStringIncludes(html, "progress-primary");
  assertStringIncludes(html, 'value="25"');
  assertStringIncludes(html, "25%");
});

Deno.test("Progress - large error progress", () => {
  const html = renderToString(Progress({
    size: "lg",
    color: "error",
    value: 90,
    striped: true,
  }));
  assertStringIncludes(html, "progress-lg");
  assertStringIncludes(html, "progress-error");
  assertStringIncludes(html, "progress-striped");
  assertStringIncludes(html, 'value="90"');
});

Deno.test("Progress - secondary animated progress", () => {
  const html = renderToString(Progress({
    color: "secondary",
    value: 45,
    animated: true,
  }));
  assertStringIncludes(html, "progress-secondary");
  assertStringIncludes(html, "progress-animated");
  assertStringIncludes(html, 'value="45"');
});

Deno.test("Progress - accent striped progress", () => {
  const html = renderToString(Progress({
    color: "accent",
    value: 70,
    striped: true,
  }));
  assertStringIncludes(html, "progress-accent");
  assertStringIncludes(html, "progress-striped");
  assertStringIncludes(html, 'value="70"');
});

Deno.test("Progress - info progress with custom max", () => {
  const html = renderToString(Progress({
    color: "info",
    value: 15,
    max: 50,
  }));
  assertStringIncludes(html, "progress-info");
  assertStringIncludes(html, 'value="15"');
  assertStringIncludes(html, 'max="50"');
  assertStringIncludes(html, "15%");
});

// Edge cases
Deno.test("Progress - negative value", () => {
  const html = renderToString(Progress({
    value: -10,
  }));
  assertStringIncludes(html, 'value="-10"');
  assertStringIncludes(html, "-10%");
});

Deno.test("Progress - decimal value", () => {
  const html = renderToString(Progress({
    value: 33.5,
  }));
  assertStringIncludes(html, 'value="33.5"');
  assertStringIncludes(html, "33.5%");
});

// Snapshot tests
Deno.test("Progress - HTML snapshot basic", async (t) => {
  const html = renderToString(Progress({
    value: 50,
    color: "primary",
  }));
  await assertSnapshot(t, html);
});

Deno.test("Progress - HTML snapshot indeterminate", async (t) => {
  const html = renderToString(Progress({
    indeterminate: true,
    color: "secondary",
    animated: true,
  }));
  await assertSnapshot(t, html);
});

Deno.test("Progress - HTML snapshot styled", async (t) => {
  const html = renderToString(Progress({
    value: 75,
    size: "lg",
    color: "success",
    striped: true,
    animated: true,
  }));
  await assertSnapshot(t, html);
});

Deno.test("Progress - HTML snapshot all sizes", async (t) => {
  const sizes = ["xs", "sm", "md", "lg"];
  const htmls = sizes.map((size) =>
    renderToString(Progress({
      size: size as 'xs' | 'sm' | 'md' | 'lg' | 'xl',
      value: 60,
      color: "primary",
    }))
  );
  await assertSnapshot(t, htmls.join("\n"));
});

Deno.test("Progress - HTML snapshot all colors", async (t) => {
  const colors = ["primary", "secondary", "accent", "info", "success", "warning", "error"];
  const htmls = colors.map((color) =>
    renderToString(Progress({
      color: color as 'primary' | 'secondary' | 'accent' | 'neutral' | 'info' | 'success' | 'warning' | 'error',
      value: 45,
      size: "md",
    }))
  );
  await assertSnapshot(t, htmls.join("\n"));
});
