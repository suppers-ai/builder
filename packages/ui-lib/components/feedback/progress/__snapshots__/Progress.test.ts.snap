export const snapshot = {};

snapshot[`ProgressDisplay - HTML snapshot basic 1`] = `'<progress class="progress progress-primary progress-md" value="50" max="100">50%</progress>'`;

snapshot[`ProgressDisplay - HTML snapshot indeterminate 1`] = `'<progress class="progress progress-secondary progress-md progress-animated"></progress>'`;

snapshot[`ProgressDisplay - HTML snapshot styled 1`] = `'<progress class="progress progress-success progress-lg progress-striped progress-animated" value="75" max="100">75%</progress>'`;

snapshot[`ProgressDisplay - HTML snapshot all sizes 1`] = `
'<progress class="progress progress-primary progress-xs" value="60" max="100">60%</progress>
<progress class="progress progress-primary progress-sm" value="60" max="100">60%</progress>
<progress class="progress progress-primary progress-md" value="60" max="100">60%</progress>
<progress class="progress progress-primary progress-lg" value="60" max="100">60%</progress>'
`;

snapshot[`ProgressDisplay - HTML snapshot all colors 1`] = `
'<progress class="progress progress-primary progress-md" value="45" max="100">45%</progress>
<progress class="progress progress-secondary progress-md" value="45" max="100">45%</progress>
<progress class="progress progress-accent progress-md" value="45" max="100">45%</progress>
<progress class="progress progress-info progress-md" value="45" max="100">45%</progress>
<progress class="progress progress-success progress-md" value="45" max="100">45%</progress>
<progress class="progress progress-warning progress-md" value="45" max="100">45%</progress>
<progress class="progress progress-error progress-md" value="45" max="100">45%</progress>'
`;
